### PR TITLE
Add information about Clock program into the CLI

### PR DESCRIPTION
**Pull request by pwittich**

### CODE REVIEW

File: projects/cm_mcu/LocalTasks.c [hunk 1]
```diff
@@ -583,21 +585,21 @@ int8_t getFFtemp(const uint8_t i)
...
   if (i < NFIREFLIES_IT_F1) {
     int index = i * (ffl12_f1_args.n_commands * ffl12_f1_args.n_pages) + i1;
-    val = ffl12_f1_args.sm_values[index];
+    val = (int8_t)ffl12_f1_args.sm_values[index];
   }
 
   else if (NFIREFLIES_IT_F1 <= i && i < NFIREFLIES_IT_F1 + NFIREFLIES_DAQ_F1) {
     int index = (i - NFIREFLIES_IT_F1) * (ffldaq_f1_args.n_commands * ffldaq_f1_args.n_pages) + i1;
-    val = ffldaq_f1_args.sm_values[index];
+    val = (int8_t)ffldaq_f1_args.sm_values[index];
```
Comment (pkotamnives) [inline projects/cm_mcu/LocalTasks.c:593]: will have to check if it's (int16_t) or (int8_t).

**Comment by pkotamnives** [comment]
@pwittich , thanks for adding this clock program info to `clkmon` in CLI. I have one comment for now after looking at the output on the FPGA-less board. I think the program name for a clock that has not yet been programed should be "notfound" or something. Currently, it gives a garbage string from clkmon.

File: projects/cm_mcu/LocalTasks.c (hunk 1 above)
Comment (pwittich) [reply projects/cm_mcu/LocalTasks.c:593]: The value printed out and the return value of the function is `int8_t` so that is why I used that for the cast.

**Comment by pwittich** [comment]
> @pwittich , thanks for adding this clock program info to `clkmon` in CLI. I have one comment for now after looking at the output on the FPGA-less board. I think the program name for a clock that has not yet been programed should be "notfound" or something. Currently, it gives a garbage string from clkmon.

Yeah we can do that. But how do we know if the clock chips has not been programmed?

**Comment by pwittich** [comment]
FWIW, I looked into the codacy warning and it is a false positive about using a VLA array. The tool doesn't find the right header file and therefore doesn't recognize that the array is in fact statically sized.

File: projects/cm_mcu/LocalTasks.c (hunk 1 above)
Comment (pkotamnives) [reply projects/cm_mcu/LocalTasks.c:593]: Okay, I see your point. I was looking at the `sm_values`'s type.

**Comment by pkotamnives** [comment]
> > @pwittich , thanks for adding this clock program info to `clkmon` in CLI. I have one comment for now after looking at the output on the FPGA-less board. I think the program name for a clock that has not yet been programed should be "notfound" or something. Currently, it gives a garbage string from clkmon.
> 
> Yeah we can do that. But how do we know if the clock chips has not been programmed?

I can check a set of three registers by clock id as I mentioned in the meeting for this. I will push this change to this PR for you to review soon.

**Comment by pwittich** [comment]
> > > @pwittich , thanks for adding this clock program info to `clkmon` in CLI. I have one comment for now after looking at the output on the FPGA-less board. I think the program name for a clock that has not yet been programed should be "notfound" or something. Currently, it gives a garbage string from clkmon.
> > 
> > 
> > Yeah we can do that. But how do we know if the clock chips has not been programmed?
> 
> I can check a set of three registers by clock id as I mentioned in the meeting for this. I will push this change to this PR for you to review soon.

Are you checkin the EEPROM? But what if the clocks were programmed by a python script?

**Comment by pkotamnives** [comment]
> > > > @pwittich , thanks for adding this clock program info to `clkmon` in CLI. I have one comment for now after looking at the output on the FPGA-less board. I think the program name for a clock that has not yet been programed should be "notfound" or something. Currently, it gives a garbage string from clkmon.
> > > 
> > > 
> > > Yeah we can do that. But how do we know if the clock chips has not been programmed?
> > 
> > 
> > I can check a set of three registers by clock id as I mentioned in the meeting for this. I will push this change to this PR for you to review soon.
> 
> Are you checkin the EEPROM? But what if the clocks were programmed by a python script?

Yes, that is what I did. I will have to think more about the later possibility. Maybe there is readonly register to check for a clock status on being programmed or not.

**Comment by pwittich** [comment]
> Yes, that is what I did. I will have to think more about the later possibility. Maybe there is readonly register to check for a clock status on being programmed or not.

Can we just add a separate printout to the clockmon command that tells if there is information in the EEPROM for that clock? I would like to keep a clean separation between inference (“the clock wasn’t programmed because there is nothing in the EEPROM”) and actual facts (“the clock has nothing in these registers, and the eeprom doesn’t have anything in these registers”)

**Comment by pkotamnives** [comment]
> > Yes, that is what I did. I will have to think more about the later possibility. Maybe there is readonly register to check for a clock status on being programmed or not.
> 
> Can we just add a separate printout to the clockmon command that tells if there is information in the EEPROM for that clock? I would like to keep a clean separation between inference (“the clock wasn’t programmed because there is nothing in the EEPROM”) and actual facts (“the clock has nothing in these registers, and the eeprom doesn’t have anything in these registers”)

Sure, that works too. I will implement that.

**Comment by pkotamnives** [comment]
@pwittich, I am not sure how to fix the Codacy issue with a VLA array that might be a false positive? but the implementation for showing clock programs in both the clock device and eeprom has been complete.

**Comment by pwittich** [comment]
> @pwittich, I am not sure how to fix the Codacy issue with a VLA array that might be a false positive? but the implementation for showing clock programs in both the clock device and eeprom has been complete.

It's a bogus error we should ignore.

**Comment by pwittich** [comment]
@pkotamnives if I read out clockmon on the board w/o FPGAs in PSB375 I find
```
% clkmon 3
Monitoring SI clock with id R1B
REG_TABLE       REG_ADDR BIT_MASK  VALUE 
PN_BASE         : 0x0002   0xff    0x0095
PN_BASE         : 0x0003   0xff    0x0053
DEVICE_REV      : 0x0005   0xff    0x0000
I2C_ADDR        : 0x000b   0x7f    0x0068
LOSXAXB         : 0x000c   0x02    0x0000
LOSOFF_IN       : 0x000d   0xff    0x0000
LOL             : 0x000e   0x02    0x0002

Program (read from clock chip): R1Bv0�03
Program (read from eeprom): R1Bv0003
```
Notice that the data read from the clock chip is corrupted. This happens on every boot. However, the EEPROM printout suggests that it is ok in the EEPROM. How is this possible?

**Comment by pkotamnives** [comment]
> @pkotamnives if I read out clockmon on the board w/o FPGAs in PSB375 I find
> 
> ```
> % clkmon 3
> Monitoring SI clock with id R1B
> REG_TABLE       REG_ADDR BIT_MASK  VALUE 
> PN_BASE         : 0x0002   0xff    0x0095
> PN_BASE         : 0x0003   0xff    0x0053
> DEVICE_REV      : 0x0005   0xff    0x0000
> I2C_ADDR        : 0x000b   0x7f    0x0068
> LOSXAXB         : 0x000c   0x02    0x0000
> LOSOFF_IN       : 0x000d   0xff    0x0000
> LOL             : 0x000e   0x02    0x0002
> 
> Program (read from clock chip): R1Bv0�03
> Program (read from eeprom): R1Bv0003
> ```
> 
> Notice that the data read from the clock chip is corrupted. This happens on every boot. However, the EEPROM printout suggests that it is ok in the EEPROM. How is this possible?

I just read one register from EEPROM that tells me the clock program number, and the rest of string (i.e. "v000") is fixed. For the one read from clock chip directly I suspect that the program I loaded to the clock might have some issues? I can check if it is also corrupted on other boards with this clock chip R1b programed. 

Though, I also notice that when the power is off, the DESIGN_ID register read from the clock turns to a number based on program 
rather than a string of program name.

**Comment by pwittich** [comment]
> 
> Though, I also notice that when the power is off, the DESIGN_ID register read from the clock turns to a number based on program rather than a string of program name.

I just tested `clkmon` with power off and it returns an empty string, which is what I expected. What did you observe?
```
% pwr off
% 202303160921 PWR INF PowerSupplyTask.c:263:power-down requested
202303160921 MONI2C INF MonitorI2CTask.c:133:FF12: PWR off. Disabling I2C monitoring.
202303160921 MONI2C INF MonitorI2CTask.c:133:CLKSI: PWR off. Disabling I2C monitoring.
202303160921 MONI2C INF MonitorI2CTask.c:133:FFDAQ: PWR off. Disabling I2C monitoring.
202303160921 PWR INF PowerSupplyTask.c:274:power-down completed
% clkmon 4
Monitoring SI clock with id R1C
REG_TABLE       REG_ADDR BIT_MASK  VALUE 
PN_BASE         : 0x0002   0xff    0x0095
PN_BASE         : 0x0003   0xff    0x0053
DEVICE_REV      : 0x0005   0xff    0x0000
I2C_ADDR        : 0x000b   0x7f    0x0068
LOSXAXB         : 0x000c   0x02    0x0000
LOSOFF_IN       : 0x000d   0xff    0x0000
LOL             : 0x000e   0x02    0x0002

Program (read from clock chip): 
Program (read from eeprom): R1Cv000X
%

File: projects/cm_mcu/clocksynth.c [hunk 2]
```diff
@@ -153,3 +154,92 @@ int load_clock()
...
+    uint32_t eepromdata[2];
+    if (PreambleList_row == 0xff && RegisterList_row == 0xffff && PostambleList_row == 0xff) { // check if a clock has been programmed or not from a set of three registers
+      eepromdata[0] = 0x58;                                                                    // supposed to be an "X" for an unprogrammed clock
+      eepromdata[1] = 0x00;
+    }
+    else {
+      eepromdata[0] = 0UL;
+      eepromdata[1] = 0UL;
+      uint32_t tempdata[2];
+      for (int i = 0; i < 4; ++i) {
```
Comment (pwittich) [inline projects/cm_mcu/clocksynth.c:223]: if we are reading 6 memory locations of a byte each why do we need 8 I2C transactions?

File: projects/cm_mcu/clocksynth.c [hunk 3]
```diff
@@ -153,3 +154,92 @@ int load_clock()
...
+    // set the page
+    uint8_t page = (CLOCK_PROGNAME_REG_ADDR_START >> 8) & 0xFF;
+    status = apollo_i2c_ctl_reg_w(CLOCK_I2C_DEV, dev_addr, 1,
+                                  CLOCK_CHANGEPAGE_REG_ADDR, 1, page);
+
+    // now read out the six bytes of data in two reads
+    const uint8_t reg = (CLOCK_PROGNAME_REG_ADDR_START)&0xFF;
+    uint16_t init_postamble_page = 32 * (device + 1) - 1;
+
+    uint32_t PreambleList_row; // the size of preamble list in a clock config file store at the end of the last eeprom page of a clock
```
Comment (pwittich) [inline projects/cm_mcu/clocksynth.c:207]: can you add some comments here what this is doing? and pls keep the lines within 80 comments

File: projects/cm_mcu/clocksynth.c (hunk 2 above)
Comment (pkotamnives) [reply projects/cm_mcu/clocksynth.c:223]: I maybe able to read one I2C transaction with 12 bytes and apply a bit mask for eeprom[0]([1]) but that is more than the 4-byte limit.

File: projects/cm_mcu/clocksynth.c (hunk 3 above)
Comment (pkotamnives) [reply projects/cm_mcu/clocksynth.c:207]: got it. I will add a descriptive comment for this.

Comment (pkotamnives) [reply projects/cm_mcu/clocksynth.c]: Done

File: projects/cm_mcu/clocksynth.c (hunk 2 above)
Comment (pwittich) [reply projects/cm_mcu/clocksynth.c:226]: that makes sense now, thanks!

**Pull request merged**